adagrams/game.py

```diff
@@ -1,11 +1,64 @@
+from random import sample
 def draw_letters():
-    pass
+    hand = sample(['A','B','C','D','E','F','G','H','I','J','K','L',\
+                   'M','N','O','P','Q','R','S','T','U','V','W','X',\
+                   'Y','Z'], counts = [9,2,2,4,12,2,3,2,9,1,1,4,2,6,\
+                                       8,2,1,6,4,6,4,2,2,1,2,1], k = 10)
+    return hand
 
 def uses_available_letters(word, letter_bank):
-    pass
+    letter_bank_dict = {}
+    for letter in letter_bank:
+        if letter not in letter_bank_dict:
+            letter_bank_dict[letter] = 1
+        else:
+            letter_bank_dict[letter] += 1
+    for letter in word.upper():
+        if letter not in letter_bank_dict:
+            return False
+        elif letter_bank_dict[letter] == 0:
+            return False
+        else:
+            letter_bank_dict[letter] -= 1
+    return True
+
 
 def score_word(word):
-    pass
+    scores = {
+        "AEIOULNRST" : 1,
+        "DG" : 2,
+        "BCMP" : 3,
+        "FHVWY" : 4,
+        "K" : 5,
+        "JX" : 8,
+        "QZ" : 10
+    }
+
+    score = [val for letter in word.upper() for key, val in scores.items() if letter in key]
+
+    if len(score) >= 7:
+        score.append(8)
+
+    return sum(score)
 
 def get_highest_word_score(word_list):
-    pass
\ No newline at end of file
+    score_list = [(word, score_word(word)) for word in word_list]
+
+    highest_scoring = [score_list[0]]
+
+    for word in score_list[1:]:
+        if word[1] == highest_scoring[0][1]:
+            highest_scoring.append(word)
+        elif word[1] > highest_scoring[0][1]:
+            highest_scoring.clear()
+            highest_scoring.append(word)
+
+    if len(highest_scoring) == 1:
+        return highest_scoring[0]
+
+    highest_scoring.sort(key=lambda word: len(word[0]))
```

Comment:
So while this is a great and correct use of `sort`, note that `sort` doesn't come for free! Python's implementation of `sort` takes O(n * log(n)) time. While most likely there'll be a lot less words in `highest_scoring` than `word_list`, if every word in `word_list` has the same score, then they will all be in `highest_score`, bringing the total time complexity here to O(n * log(n)), where n is the number of words in `word_list`. However it can be done in O(n) by not bothering to keep track of previous highest scoring words and performing the tie-breaking within the for above loop (I'm also going to make a few other nice code changes that I'll point out in comments in the snippet):

```python
for word_score_tuple in score_list:
    # using tuple destructuring to give the tuple elements some nicer names
    word, score = word_score_tuple
    highest_word, highest_score = highest_scoring
    if score == highest_score:
        highest_len = len(highest_word) # only using a variable here and with word_len for a nicer name
        if highest_len != 10:
            word_len = len(word)
            if word_len == 10 or word_len < highest_len:
                 highest_scoring = word_score_tuple
    elif score > highest_score:
         highest_scoring = word_score_tuple

return highest_scoring
```

Yikes! That is a lot of nesting, which hurts readability, but with the benefit that we only have to go through `score_list` once! That's O(n), which is a big improvement over O(n * log(n))!